src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
                 {
+                    // Fall through to the last available token expected in a member name for calculating the expected
+                    // position of the lParen. This accounts for things like `[Array]::Empty[int[]()` so that we don't
+                    // get unhandled behaviour when input is expected but incomplete.
+                    int endColumnNumber = rBracket?.Extent?.EndColumnNumber
+                        ?? genericTypeArguments?.LastOrDefault()?.Extent?.EndColumnNumber
+                        ?? member.Extent.EndColumnNumber;
+
                     Diagnostics.Assert(lParen.Kind == TokenKind.LParen || lParen.Kind == TokenKind.LCurly, "token kind incorrect");
-                    Diagnostics.Assert(member.Extent.EndColumnNumber == lParen.Extent.StartColumnNumber,
-                                       "member and paren must be adjacent");
-                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member);
+                    Diagnostics.Assert(
+                        endColumnNumber == lParen.Extent.StartColumnNumber,
+                        "member and paren must be adjacent when the method is not generic");
+                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member, genericTypeArguments);
                 }
             }
 
             return new MemberExpressionAst(
-                    ExtentOf(targetExpr, member),
-                    targetExpr,
-                    member,
-                    @static: operatorToken.Kind == TokenKind.ColonColon,
-                    nullConditional: operatorToken.Kind == TokenKind.QuestionDot);
+                ExtentOf(targetExpr, member),
+                targetExpr,
+                member,
+                @static: operatorToken.Kind == TokenKind.ColonColon,
+                nullConditional: operatorToken.Kind == TokenKind.QuestionDot,
+                genericTypeArguments);
+        }
+
+        private List<ITypeName> GenericMethodTypeArgumentsRule(out Token rBracketToken)
+        {
+            List<ITypeName> genericTypes = null;
+
+            int resyncIndex = _tokenizer.GetRestorePoint();
+            Token lBracket = NextToken();
+            rBracketToken = null;
+
+            if (lBracket.Kind == TokenKind.LBracket)
+            {
+                // This is either a member expression with generic type arguments, or some sort of collection index
+                // on a property.
+                var oldTokenizerMode = _tokenizer.Mode;
+                try
+                {
+                    // Switch to typename mode to avoid aggressive argument tokenization.
+                    SetTokenizerMode(TokenizerMode.TypeName);
+
+                    SkipNewlines();
+                    Token firstToken = NextToken();
+                    if (firstToken.Kind != TokenKind.Number
+                        && (firstToken.Kind == TokenKind.Identifier || firstToken.Kind == TokenKind.LBracket))
```

Comment:
The logic here seems to test that the token kind is not one thing and then is one of two others. The second conditions should preclude the first, no?

One question is what unexpected tokens do here and how we signal that to the caller

Reply:
If we can identify some variations on the logic here, we should ideally add tests for each possibility so that we have a strong understanding of the behaviour here

Reply:
So I could be misremembering, but I think `Identifier` _includes_ `Number` (it's a named flag combination, if I recall correctly?)

Will have to look back over this code to answer your other point. 🙂

Reply:
Re-checked this, and yeah you're right, the first condition isn't needed. I must have been thinking of something else, no [Flags] shenanigans here. ^^

As for what unexpected tokens do here.... not sure exactly what you mean. Looking at the code, we use ReportIncompleteInput() but beyond that I'm not quite sure. 🙂